Microsoft.Toolkit/Extensions/TypeExtensions.cs

```diff
@@ -56,70 +57,146 @@ public static class TypeExtensions
+                        var formattedTypes = type.GetGenericArguments().Select(t => FormatDisplayString(t, 0, t.GetGenericArguments()));
+
+                        return $"({string.Join(", ", formattedTypes)})";
                     }
+                }
+
+                string displayName;
+
+                // Generic types
+                if (type.Name.Contains('`'))
+                {
+                    // Retrieve the current generic arguments for the current type (leaf or not)
+                    var tokens = type.Name.Split('`');
+                    var genericArgumentsCount = int.Parse(tokens[1]);
+                    var typeArgumentsOffset = typeArguments.Length - genericTypeOffset - genericArgumentsCount;
+                    var currentTypeArguments = typeArguments.Slice(typeArgumentsOffset, genericArgumentsCount).ToArray();
+                    var formattedTypes = currentTypeArguments.Select(t => FormatDisplayString(t, 0, t.GetGenericArguments()));
 
                     // Standard generic types are displayed as Foo<T>
-                    return $"{genericName}<{string.Join(", ", typeArguments)}>";
+                    displayName = $"{tokens[0]}<{string.Join(", ", formattedTypes)}>";
+
+                    // Track the current offset for the shared generic arguments list
+                    genericTypeOffset += genericArgumentsCount;
+                }
+                else
+                {
+                    // Simple custom types
+                    displayName = type.Name;
                 }
 
-                // Array types are displayed as Foo[]
-                if (type.IsArray)
+                // If the type is nested, recursively format the hierarchy as well
+                if (type.IsNested)
                 {
-                    var elementType = type.GetElementType();
-                    var rank = type.GetArrayRank();
+                    var openDeclaringType = type.DeclaringType!;
+                    var rootGenericArguments = typeArguments.Slice(0, typeArguments.Length - genericTypeOffset).ToArray();
+
+                    // If the declaring type is generic, we need to reconstruct the closed type
+                    // manually, as the declaring type instance doesn't retain type information.
+                    if (rootGenericArguments.Length > 0)
+                    {
+                        var closedDeclaringType = openDeclaringType.GetGenericTypeDefinition().MakeGenericType(rootGenericArguments);
 
-                    return $"{FormatDisplayString(elementType)}[{new string(',', rank - 1)}]";
+                        return $"{FormatDisplayString(closedDeclaringType, genericTypeOffset, typeArguments)}.{displayName}";
+                    }
+
+                    return $"{FormatDisplayString(openDeclaringType, genericTypeOffset, typeArguments)}.{displayName}";
                 }
 
-                return type.ToString();
+                return $"{type.Namespace}.{displayName}";
```

Comment:
Is the output here significantly different or just better formatted here? i.e. if someone with 6.1 upgrades, will the result be different in the cases where things worked?

(Just want to check if we should mark this as a breaking change even though the API surface itself didn't change.)

Reply:
The output should be the same for types that were properly supported already 😊
In that line I'm explicitly adding the namespace just because with this change we're now using the name of each type directly instead of the fullname (so that we can manually traverse each type in case of nested types), so we need to add the namespace at the start ourselves to match the previous output with the fully qualified name.

As for changes between this PR and the previous version, here's some examples.
I'm only displaying cases with actual differences, such as nested types or pointer/ref types:

```csharp
typeof(Animal.Cat)
// 6.1: Namespace.Animal+Cat    (uses + instead of .)
// 7.0: Namespace.Animal.Cat

typeof(Animal.Cat<int>)
// 6.1: Namespace.Animal+Cat<int>    (same)
// 7.0: Namespace.Animal.Cat<int>

typeof(Animal.Cat<int>.Bar)
// 6.1: Namespace.Animal+Cat<int>    (the nested .Bar type got lost)
// 7.0: Namespace.Animal.Cat<int>.Bar

typeof(Animal.Llama<string, int[]>.Foo)
// 6.1: Namespace.Animal+Llama<string, int[]>    (same)
// 7.0: Namespace.Animal.Llama<string, int[]>.Foo

typeof(Animal.Llama<string, int[]>.Foo<byte>)
// 6.1: Namespace.Animal+Llama<string, int[], byte>    (nested type missing, wrong type arguments)
// 7.0: Namespace.Animal.Llama<string, int[]>.Foo<byte>

typeof(void)
// 6.1: System.Void
// 7.0: void

typeof(int**)
// 6.1: System.Int32**
// 7.0: int**

typeof(string).MakeByRefType()
// 6.1: System.String&
// 7.0: string&
```

In general, some cases just look nicer (eg. `.` instead of `+` for a simple nested type, or `void` instead of `System.Void`), while others are now actually correct, whereas currently they are either missing bits (eg. nested types getting lost in the process) and/or displayed with wrong type arguments (they're basically all combined together).